packages/apputils/src/toolbar/factory.ts

```diff
@@ -0,0 +1,270 @@
+    // prompt the user to reload if the toolbar definition has been updated.
+    const newItems = (settings.composite[propertyId] as any) ?? [];
+    if (!JSONExt.deepEqual(toArray(toolbarItems.iter()), newItems)) {
+      void displayInformation(trans);
+    }
+  });
+
+  // React to plugin changes
+  registry.pluginChanged.connect(async (sender, plugin) => {
+    if (plugin !== pluginId) {
+      // If a plugin changed its toolbar items
+      const oldItems = loaded[plugin] ?? [];
+      const newItems =
+        (registry.plugins[plugin]!.schema['jupyter.lab.toolbars'] ?? {})[
+          factoryName
+        ] ?? [];
+      if (!JSONExt.deepEqual(oldItems, newItems)) {
+        if (loaded[plugin]) {
+          // The plugin has changed, request the user to reload the UI
+          await displayInformation(trans);
+        } else {
+          // The plugin was not yet loaded => update the toolbar items list
+          loaded[plugin] = JSONExt.deepCopy(newItems);
+          const newList =
+            SettingRegistry.reconcileToolbarItems(
+              toArray(toolbarItems),
+              newItems,
+              false
+            ) ?? [];
+
+          // Existing items cannot be removed.
+          newList?.forEach(item => {
+            const index = findIndex(
+              toolbarItems,
+              value => item.name === value.name
+            );
+            if (index < 0) {
+              toolbarItems.push(item);
+            } else {
+              toolbarItems.set(index, item);
+            }
+          });
+        }
+      }
+    }
+  });
+
+  return toolbarItems;
+}
+
+/**
+ * Create the toolbar factory for a given container widget based
+ * on a data description stored in settings
+ *
+ * @param toolbarRegistry Toolbar widgets registry
+ * @param settingsRegistry Settings registry
+ * @param factoryName Toolbar container factory name
+ * @param pluginId Settings plugin id
+ * @param translator Translator
+ * @param propertyId Toolbar definition key in the settings plugin
+ * @returns List of toolbar widgets
+ */
+export function createToolbarFactory(
+  toolbarRegistry: IToolbarWidgetRegistry,
+  settingsRegistry: ISettingRegistry,
+  factoryName: string,
+  pluginId: string,
+  translator: ITranslator,
+  propertyId: string = 'toolbar'
+): (widget: Widget) => ToolbarRegistry.IToolbarItem[] {
+  const items: ToolbarRegistry.IWidget[] = [];
+  let rawItems: IObservableList<ToolbarRegistry.IWidget>;
+
+  const transfer = (
+    list: IObservableList<ToolbarRegistry.IWidget>,
+    change: IObservableList.IChangedArgs<ToolbarRegistry.IWidget>
+  ) => {
+    switch (change.type) {
+      case 'move':
+        break;
+      case 'add':
+      case 'remove':
+      case 'set':
+        items.length = 0;
+        items.push(
+          ...toArray(list)
+            .filter(item => !item.disabled)
+            .sort((a, b) => (a.rank ?? Infinity) - (b.rank ?? Infinity))
+        );
```

Comment:
I'm not following the logic here. It looks like action for all types of change is the same except for move. Why is that?

Reply:
Plugin may register their settings lately. So the toolbar `items` are by default empty and they are updated by observing the `rawItems` list. That list is initialized when the settings registry is available and when a new plugin get its settings loaded.

The observable list has four change types. All will impact the toolbar items list (aka clear it and copy the enabled ones from the settings) except the `move` change as the position of the items is set uniquely by the item rank. Hence we don't do anything for `move`.

Reply:
Thanks!